src/firstboot/firstboot.c

```diff
@@ -458,8 +458,20 @@ static int process_locale(int rfd) {
         return 1;
 }
 
+static bool keymap_exists_bool(const char *name) {
+        return keymap_exists(name) > 0;
+}
+
+static bool keymap_is_ok(int rfd, const char *name) {
+        assert(rfd >= 0);
+
+        return dir_fd_is_root(rfd) ? keymap_exists_bool(name) : keymap_is_valid(name);
+}
+
 static int prompt_keymap(int rfd) {
         _cleanup_strv_free_ char **kmaps = NULL;
+        bool (*is_valid)(const char *name) = dir_fd_is_root(rfd) ? keymap_exists_bool
```

Comment:
this ternary op you basically have in place twice. You could move the ternary op that determines the function into a function itself... i.e.

```c
static typeof(keymap_is_valid) determine_keymap_validity_func(int rfd) {
       int r;

       r = dir_fd_is_root(rfd);
       if (r < 0) 
               log_debug_errno(r, "Unable to determine if operating on host root directory, assuming we are: %m");

       return r != 0 ? keymap_exists_bool : keymap_is_valid;
}
```

Reply:
To get it to build I had to do `static typeof(&keymap_is_valid)`, which I think makes sense as we are returning a function pointer - alternatively the normal ugly C fptr syntax `static bool (*determine_keymap_validity_func(int rfd))(const char*)` works fine